internal/rbd/controllerserver.go

```diff
@@ -1456,13 +1470,21 @@ func (cs *ControllerServer) ControllerExpandVolume(
 	}
 	defer cs.VolumeLocks.Release(volID)
 
-	cr, err := util.NewUserCredentials(req.GetSecrets())
+	secrets := req.GetSecrets()
+	if util.IsMigrationSecret(secrets) {
+		secrets, err = util.ParseAndSetSecretMapFromMigSecret(secrets)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+	}
+
+	cr, err := util.NewUserCredentials(secrets)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	defer cr.DeleteCredentials()
 
-	rbdVol, err := genVolFromVolID(ctx, volID, cr, req.GetSecrets())
+	rbdVol, err := genVolFromVolID(ctx, volID, cr, secrets)
```

Comment:
These `secrets` are only needed for KMS (encryption passphrase in the linked StorageClass Secret), so the migration secrets (cluster access credentials) are not needed.

```go
	secrets := req.GetSecrets()
	if util.IsMigrationSecret(secrets) {
		secrets, err = util.ParseAndSetSecretMapFromMigSecret(secrets)
		if err != nil {
			return nil, status.Error(codes.InvalidArgument, err.Error())
		}
	}

	cr, err := util.NewUserCredentials(secrets)
```

can therefore be replaced by a single  helper like

```go
cr, err := util.NewUserCredentialsWithMigration(req.GetSecrets());
```

or something along those lines. When the helper does not detect `IsMigrationSecret()`, it can return `NewUserCredentials()` directly. Such a single helper is much cleaner and easier to maintain copy/pasted code in different locations. It also reduces the :spaghetti: code sprinkled over multiple places for migration support.

Reply:
Sure, @nixpanic , will adopt that model here.

Reply:
Done.